### PR TITLE
feat: reuse HTTP sessions

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

Introduce the option of using persistent HTTP sessions in `api.Session`.

Previously, each API call was wrapped in its own `requests.Session`, which meant making a new underlying TCP connection/TLS handshake every time. Sometimes this is necessary because we can't know when to close the session, and we don't want to leave too many dangling HTTP connections open. But in many cases, we make lots of API calls in quick succession within known bounds, and in these cases using a persistent HTTP session significantly improves performance:

**Reporting metrics in Core API/Detached mode:**
```python
def main():
    core_v2.init(
        defaults=core_v2.DefaultConfig(
            name="unmanaged-1-singleton",
        ),
    )

    for i in range(100):
        core_v2.train.report_training_metrics(
            steps_completed=i, metrics={"loss": random.random()}
        )

        if (i + 1) % 10 == 0:
            core_v2.train.report_validation_metrics(
                steps_completed=i, metrics={"loss": random.random()}
            )

    core_v2.close()
```
*Not using persistent HTTP sessions:* 13-14s
*With a persistent HTTP session:* 6-7s

**CLI calls that fetch multiple pages:**
i.e. `list-experiments`, `list-checkpoints`
*Before:* 0.75s
*After:* 0.3s

### `api.Session` as a context manager

This PR exposes the option of re-using the same underlying HTTP session through `api.Session` as a context manager:

```python
with api.Session(...) as sess:
  # All requests within the context share the same persistent HTTP connection.
  # Once the context exits, the session is closed.
  sess.post(...)
  sess.get(...)
```

When used directly, each HTTP call will create a new session every time. This is the current behavior, so existing code will continue to function the same way:

```python
# When called directly outside of a context, each request uses a new HTTP session.
sess = api.Session(...)
sess.post(...)
sess.get(...)
```

This functionality is implemented in Core API, where all internal API calls within a `core.Context` (`with core.init(...) as core_context:`) will now share the same persistent HTTP connection which is closed when the `core_context` context exits.

#### CLI

In the CLI, `api.Session` as a context manager is implemented as a decorator, and when used, exposes a session that is persistent within the decorated function:

```diff
- def list_experiments(args: Namespace) -> None:
-     sess = cli.setup_session(args)

+ @cli.session
+ def list_experiments(args: Namespace, sess: api.Session) -> None:
```

I only changed a few select APIs/CLIs to use the new functionality, mostly as examples of how it would be used. There are many places that would see performance benefits from persistent sessions, but I imagine that will probably be a separate effort/PR.

### Other `api.Session`refactors
Some other small refactors made to `api.Session`:
- Set auth/cert-related parameters on the session instead of passing them into each request
- `harness/determined/common/requests.py` was deleted, its logic consolidated into `harness/determined/common/api/_session.py`

## Test Plan

- Run any experiment that reports a bunch of metrics (ideally Core API / detached mode) and enable debug logs.
- Submit the experiment and the logs should show "urllib3.connectionpool: Starting new HTTP connection" only once at the start of training
- Experiment should be noticeably faster

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/cli/_util.py [hunk 1]
```diff
@@ -99,6 +100,22 @@ def setup_session(args: argparse.Namespace) -> api.Session:
     )
 
 
+def session(
+    fn: Callable[[argparse.Namespace, api.Session], None]
+) -> Callable[[argparse.Namespace, api.Session], None]:
```
Comment (ioga) [inline harness/determined/cli/_util.py:105]:
```suggestion
) -> Callable[argparse.Namespace, None]:
```

cause the wrapped function does not take 2nd `api.Session` param anymore?

File: harness/determined/common/api/_session.py [hunk 2]
```diff
@@ -174,9 +224,11 @@ def put(
...
     def with_retry(self: T, max_retries: GeneralizedRetry) -> T:
         """Generate a new session with a different retry policy."""
-        new_session = copy.copy(self)
-        new_session._max_retries = max_retries
-        return new_session
+        self._max_retries = max_retries
+        if self._http_session:
+            self._http_session.close()
+            self._http_session = self._make_http_session()
+        return self
```
Comment (ioga) [inline harness/determined/common/api/_session.py:231]: this changes the semantics a bit. I think the old method and `with_` naming and its docstring indicated you make a new session like the old one but with this one setting different. you can continue using the old session object with the old setting.

now, the underlying object is updated in place with the new setting. it does not seem like this method is ever used in a way where the old object is still preserved. so the change in semantics is fine. but maybe 
1. update the docstring to reflect the different behavior
2. consider renaming it to `set_retry` or something to reflect it updates the thing itself

File: harness/tests/common/api/test_session.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        num_requests = 5
+
+        # Make a few requests using the same session object.
+        for i in range(num_requests):
+            resp = session.get(
+                path="/info",
+                params=None,
+                headers=None,
+                timeout=None,
+                stream=False,
```
Comment (ioga) [inline harness/tests/common/api/test_session.py:25]: aren't all these parameters redundant?

File: harness/determined/cli/_util.py (hunk 1 above)
Comment (azhou-determined) [reply harness/determined/cli/_util.py]: done

**Review by ioga — APPROVED** [review]

File: harness/determined/cli/_util.py [hunk 4]
```diff
@@ -99,6 +100,22 @@ def setup_session(args: argparse.Namespace) -> api.Session:
     )
 
 
+def session(
```
Comment (wes-turner) [inline harness/determined/cli/_util.py:103]: If this decorator is a necessary part of having persistent HTTP sessions, I haven't figured out the mechanism.

File: harness/determined/common/api/_session.py [hunk 5]
```diff
@@ -22,69 +22,22 @@
...
+    )
+    requests_session.mount("http://", adapters.HTTPAdapter(max_retries=max_retries))
+    requests_session.verify = verify
+    if headers:
+        requests_session.headers.update(headers)
+
+    return requests_session
 
 
 class BaseSession(metaclass=abc.ABCMeta):
```
Comment (wes-turner) [inline harness/determined/common/api/_session.py:43]: I think it makes sense for a Determined `Session` to follow the same interface and have the same semantics as a requests `Session` (maybe more accurately to implement a subset of that functioanlity). Requests offers the direct interface as well as a context manager, and it keeps connections open in both of them. (https://requests.readthedocs.io/en/latest/user/advanced/#session-objects)

Comment (wes-turner) [reply harness/determined/common/api/_session.py:43]: A long discussion reached the decision to keep the differing behavior as implemented where when used as a context manager a Determined Session uses persistent HTTP sessions and when assigned to a variable directly it uses 1-time HTTP sessions. This may be changed at a later date.

File: harness/determined/common/api/_util.py [hunk 6]
```diff
@@ -110,6 +111,26 @@ def read_paginated(
         offset = pagination.endIndex
 
 
+def read_paginated_with_session(
```
Comment (wes-turner) [inline harness/determined/common/api/_util.py:114]: suggestion: add an optional `session` to `read_paginated`?

File: harness/determined/common/experimental/experiment.py [hunk 7]
```diff
@@ -514,10 +516,12 @@ def get_with_offset(offset: int) -> bindings.v1GetExperimentCheckpointsResponse:
                 states=[bindings.checkpointv1State.COMPLETED],
             )
 
-        resps = api.read_paginated(
-            get_with_offset=get_with_offset,
-            pages=api.PageOpts.single if max_results else api.PageOpts.all,
-        )
+        with self._session as sess:
+            resps = api.read_paginated(
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:520]: This is still one of my least-favorite parts of the SDK. Would a future design want to build this into `BaseSession`?

File: harness/determined/common/experimental/experiment.py [hunk 8]
```diff
@@ -498,9 +498,11 @@ def list_checkpoints(
         if not sort_by:
             sort_by = checkpoint.CheckpointSortBy.SEARCHER_METRIC
 
-        def get_with_offset(offset: int) -> bindings.v1GetExperimentCheckpointsResponse:
+        def get_with_offset(
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:501]: Could you please explain to me why this change to `get_with_offset` was necessary?

File: harness/determined/common/experimental/experiment.py [hunk 9]
```diff
@@ -514,10 +516,12 @@ def get_with_offset(offset: int) -> bindings.v1GetExperimentCheckpointsResponse:
...
 
-        resps = api.read_paginated(
-            get_with_offset=get_with_offset,
-            pages=api.PageOpts.single if max_results else api.PageOpts.all,
-        )
+        with self._session as sess:
+            resps = api.read_paginated(
+                get_with_offset=get_with_offset,
+                pages=api.PageOpts.single if max_results else api.PageOpts.all,
+                session=sess,
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:523]: As currently written with separate `read_paginated` and `read_paginated_with_session`, I don't understand how this works. `session` isn't an argument of `read_paginated`.

File: harness/determined/common/api/_util.py (hunk 6 above)
Comment (wes-turner) [reply harness/determined/common/api/_util.py:114]: Actually... I don't get why this was necessary to do at all. `get_with_offset` has a `session` already referenced. Why would we need do give it a different one rather than the one that tends to be in scope when it's defined?

File: harness/determined/common/api/_session.py [hunk 10]
```diff
@@ -95,13 +48,59 @@ class BaseSession(metaclass=abc.ABCMeta):
...
+    _http_session: Optional[requests.Session]
+
+    def __enter__(self):
+        self.persist_http_session()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def persist_http_session(self) -> None:
```
Comment (wes-turner) [inline harness/determined/common/api/_session.py:88]: If this is meant to be a public method (or at least public within the repo), please give these docstrings.

File: harness/determined/common/api/_session.py [hunk 11]
```diff
@@ -113,7 +112,59 @@ def _do_request(
...
+        except requests.exceptions.ConnectionError as e:
+            raise errors.MasterNotFoundException(str(e))
+        except requests.exceptions.RequestException as e:
+            raise errors.BadRequestException(str(e))
+        finally:
+            if not self._http_session:
+                # Close the session if not persistent.
+                session.close()
+
+        def _get_error_str(r: requests.models.Response) -> str:
```
Comment (wes-turner) [inline harness/determined/common/api/_session.py:147]: This already violated our style guide (https://google.github.io/styleguide/pyguide.html#264-decision), and I think it's worse now that it's in a class.

> Avoid nested functions or classes except when closing over a local value other than self or cls. Do not nest a function just to hide it from users of a module. Instead, prefix its name with an _ at the module level so that it can still be accessed by tests.

Not your job, but bonus if you'd like to un-nest it.

File: harness/determined/common/api/_session.py [hunk 12]
```diff
@@ -22,69 +22,22 @@
...
-        raise errors.ForbiddenException(message=_get_error_str(r))
-    if r.status_code == 401:
-        raise errors.UnauthenticatedException()
-    elif r.status_code == 404:
-        raise errors.NotFoundException(_get_error_str(r))
-    elif r.status_code >= 300:
-        raise errors.APIException(r)
-
-    return r
+def _make_requests_session(
```
Comment (wes-turner) [inline harness/determined/common/api/_session.py:25]: What difference should the reader make between a `requests_session` and an `http_session`?

File: harness/determined/common/api/_session.py [hunk 13]
```diff
@@ -22,69 +22,22 @@
...
+    server_hostname: Optional[str] = None,
+    verify: Optional[Union[str, bool]] = None,
+    max_retries: Optional[GeneralizedRetry] = None,
+    headers: Optional[Dict[str, Any]] = None,
+) -> requests.Session:
+    requests_session = requests.Session()
+    requests_session.mount(
+        "https://", _HTTPSAdapter(server_hostname=server_hostname, max_retries=max_retries)
+    )
+    requests_session.mount("http://", adapters.HTTPAdapter(max_retries=max_retries))
```
Comment (wes-turner) [inline harness/determined/common/api/_session.py:35]: The `HTTPAdapter` for `https://` has a `server_hostname` and the one for `http://` doesn't. Is that a problem?

File: harness/determined/common/api/_session.py (hunk 12 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py:25]: It seems to me like a `BaseSession._http_session` is better named a `_requests_session`, FWIW.

File: harness/determined/common/api/_session.py [hunk 14]
```diff
@@ -22,69 +22,22 @@
...
-        raise errors.UnauthenticatedException()
-    elif r.status_code == 404:
-        raise errors.NotFoundException(_get_error_str(r))
-    elif r.status_code >= 300:
-        raise errors.APIException(r)
-
-    return r
+def _make_requests_session(
+    server_hostname: Optional[str] = None,
+    verify: Optional[Union[str, bool]] = None,
```
Comment (wes-turner) [inline harness/determined/common/api/_session.py:27]: When `verify` is optional, then I have to figure out what three states mean. I think it's slightly easier (and means the same thing) if it's just a `Union[str, bool]` instead.

File: harness/determined/common/api/_session.py (hunk 13 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py:35]: NM. I get it now. Where did this pattern come from?

File: harness/determined/common/api/_session.py (hunk 12 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py:25]: ok -- suggestion: replace all the `_make_http_session` methods with calls to `_make_requests_session` that pass explicit values for all these parameters. That way I don't have to cross-reference attributes and figure out what they may have been set to from elsewhere (`with_retry` took a lot of thinking before I felt comfortable that it was correct)

Comment (wes-turner) [reply harness/determined/common/api/_session.py:25]: Well, now I see that isn't possible. This seems like a smell. Thinking about how to fix it.

Comment (wes-turner) [reply harness/determined/common/api/_session.py:25]: Well... this would be simpler if, like in requests, the resources were created during `__init__`. In requests.sessions.Session:

```python
    def __enter__(self):
        return self
```

Comment (wes-turner) [reply harness/determined/common/api/_session.py:25]: Is there any danger in doing it this way? I don't think a connection will be opened inside the requests Session until a GET, etc. is called.

File: harness/determined/core/_context.py [hunk 15]
```diff
@@ -39,6 +39,7 @@ class Context:
 
     def __init__(
         self,
+        _session: api.Session,
```
Comment (wes-turner) [inline harness/determined/core/_context.py:42]: I don't like that I don't understand how this works. I can't tell how much of that is my fault, how much the PR's. When you're ready to talk about this part, could we chat?

File: harness/determined/common/api/_session.py (hunk 13 above)
Comment (azhou-determined) [reply harness/determined/common/api/_session.py:35]: we don't need to worry about custom TLS certs if it's not HTTPS?

File: harness/determined/common/api/_session.py (hunk 14 above)
Comment (azhou-determined) [reply harness/determined/common/api/_session.py:27]: we probably could default None to True, but it's not that obvious and i'm not sure that it's this method's responsibility to make that default decision.

`verify` comes directly from `cert.bundle`, which has `Optional[Union[str, bool]]`, along with some not-super-trivial logic to determine which type it is:
```
if noverify:
            self._bundle = False  # type: Union[None, str, bool]
        elif cert_pem is None:
            self._bundle = None
        else:
            self._bundle = str
```
 requests also takes in the same type and has some logic for handling None (ultimately it treats it as true AFAICT, but it's not immediately obvious).

File: harness/determined/common/api/_util.py (hunk 6 above)
Comment (azhou-determined) [reply harness/determined/common/api/_util.py:114]: this is partially because i initially implemented it in `read_paginated` but in a later commit decided to separate it out, and in doing so, left out the code that actually mattered 🙃 i fixed it, so that now `read_paginated_with_session` is responsible for the session context, and `get_with_offset` uses the session it's given.

Comment (azhou-determined) [reply harness/determined/common/api/_util.py:114]: > Why would we need do give it a different one rather than the one that tends to be in scope when it's defined?
because the scope is different. `get_with_offset` is a singular API call that can't manage a persistent session. imagine:
```
with self._session as sess:
    def get_with_offset(int):
        return bindings.get_Trials(
            session=sess
        )
    # if we call read_paginated here, it doesn't seem right that it doesn't take in a session
    resp = read_paginated(get_with_offset)

# if we call it here outside the session context, the session context is useless
resp = read_paginated(get_with_offset)
```
it makes sense to me that `read_paginated` has the responsibility of entering/exiting the session context since it knows the "context" of requesting multiple pages in succession. then `get_with_offset` just uses the session `read_paginated_with_session` gives it. 

i agree it's not great, but given the way `read_paginated` is implemented with the separate `get_with_offset`, this is the cleanest way i could find to do it

File: harness/determined/common/api/_session.py (hunk 12 above)
Comment (azhou-determined) [reply harness/determined/common/api/_session.py:25]: well, i tried this at some point. i think the awkwardness here goes back to wanting to support the existing behavior of keeping direct calls single-use sessions. 

but it was my first instinct as well to initialize the session in the __init__. it ended up being more awkward, because then there's non-obvious logic that would be required in `_do_request` on when to close them. 

in the end i felt like it was much more explicit to create resources as the behavior dictates: only on entering a context manager do we start a persistent session, and `_do_request` will use a persistent session if it exists, otherwise it will create a one-time session that only exits for the scope of the function.

File: harness/determined/common/experimental/experiment.py (hunk 7 above)
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py]: i don't love it either, but i dunno. could also see `read_paginated` belonging somewhere in `bindings`. definitely out of scope for this PR

File: harness/determined/common/api/_session.py (hunk 10 above)
Comment (azhou-determined) [reply harness/determined/common/api/_session.py]: shouldn't be public. made it private.

File: harness/determined/common/api/_util.py (hunk 6 above)
Comment (wes-turner) [reply harness/determined/common/api/_util.py:114]: I think I need it explained to me what makes this a good idea.

The implicit-session interface to `read_paginated` looks to me like:
1. construct a method that maps an int to a list of things starting at that number
2. exercise the method to get a complete list of things

The explicit-session interface is:
1. construct a method that uses a session to map an int to a list of things starting at that number
2. If the passed Session has an existing HTTP connection, close it
3. Create a new persistent HTTP connection in the session passed to `read_paginated`
4. use the modified Session to call the master and get your complete list
5. close the persistent HTTP connection in the passed Session

Modifying the Session seems bad to me. If it turns out we need an explicit-session interface, then I think it should either rely on persistent-HTTP management from the caller _or_ create a brand-new session. Relying on the caller to manage it seems like the more natural interface from the point of view of a caller for the given signature.

Your intuition and mine differ about how 
```python
resp = read_paginated(get_with_offset)
``` 
looks without a passed session. I think `read_paginated`/`get_with_offset` interface is bad, but given the relationship between them, it feels more natural to me for the session to be implicit in `get_with_offset`. Whether that's closed with `read_paginated` or closed by the caller of `read_paginated`, I don't so much care.

I don't think it's necessary to support the use case in the example above. I don't think a caller should expect that that use case works.

File: harness/determined/common/api/_session.py (hunk 11 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py]: Thanks!

File: harness/determined/common/api/_session.py (hunk 12 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py:25]: > because then there's non-obvious logic that would be required in `_do_request` on when to close them.

Could we talk through this? I think it would be nice to know that resources are created rather than have to puzzle out context.

File: harness/determined/common/api/_session.py (hunk 13 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py:35]: I meant the pattern for setting a hostname inside the subclassed `HTTPAdapter`. I assume it's good practice, but when I went to look for examples out there on the internet to learn more about it I wasn't very successful. Where'd the pattern come from?

File: harness/determined/common/api/_session.py (hunk 14 above)
Comment (wes-turner) [reply harness/determined/common/api/_session.py:27]: I'm only now looking at how `requests.Sessions` are initialized.

...

ok. So a `requests.Session` has a `verify` attribute that's set to `True` at init but which [the docs say](https://requests.readthedocs.io/en/latest/api/#requests.Session.verify) you can falsify later on. FWIW, I don't see mention of setting it to `None`.

`requests.Session.request` _does_ take an `Optional` `verify` when constructing its request, and that logic is _awful_:

1. if `request.verify \in (True | None)`, and a bundle-string exists in an environment variable, then overwrite `request.verify` with that bundle-string
2. if now `request.verify is None`, then use Session.verify for the request
3. if `request.verify is not None`, use whatever `request.verify` now is for the request

ok. It's time to send the request. Now it's up to the Adapter to decide what to do with `verify`. Looking at [the source](https://github.com/psf/requests/blob/f1bb07d39b74d6444e333879f8b8a3d9dd4d2311/src/requests/adapters.py#L124), the comment on the base class says that `verify` should be either a string or a bool. This library is untyped so whatever, but says nothing about the possibility that it's a `None`.

I think it's a type error at this point for `verify` to be a `None` unless the adapter we subclass it with explicitly handles that case. Ours does not, but having earlier set `request.verify` to `None` manually does mean that possibility is there.

My conclusion from this is that for our use case we should not set a `requests.Session.verify` to `None`.

File: harness/determined/common/api/_util.py (hunk 6 above)
Comment (azhou-determined) [reply harness/determined/common/api/_util.py:114]: removed additional changes for `read_paginated`

File: harness/determined/common/api/_session.py (hunk 14 above)
Comment (azhou-determined) [reply harness/determined/common/api/_session.py]: defaulting this to true now

File: harness/tests/common/api/test_session.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
+from determined.common import api
+from determined.common.api import certs
+from tests.common import api_server
+
+
+def test_session_non_persistent_http_sessions() -> None:
+    with api_server.run_api_server(address=("localhost", 8080)) as master_url:
```
Comment (wes-turner) [inline harness/tests/common/api/test_session.py:7]: FWIW, I think `api_server.run_api_server` was a design mistake, and that we should either mock python endpoints or HTTP calls with `responses`. I think this would be plenty easy to do here, and would be clearer to the reader, in addition.

I don't think you _have_ to migrate, but I do think it'd be better for our test health to stop using the fake master.

Comment (azhou-determined) [reply harness/tests/common/api/test_session.py:7]: believe me i really didn't want to use this. tried both mocking endpoints and `responses`. the underlying HTTP connection pools only exist for real connections

File: harness/determined/common/experimental/experiment.py [hunk 17]
```diff
@@ -498,26 +498,28 @@ def list_checkpoints(
...
-                orderBy=order_by._to_bindings() if order_by else None,
-                sortByAttr=(
-                    sort_by._to_bindings()
-                    if isinstance(sort_by, checkpoint.CheckpointSortBy)
-                    else None
-                ),
-                sortByMetric=sort_by if isinstance(sort_by, str) else None,
-                states=[bindings.checkpointv1State.COMPLETED],
-            )
+        with self._session as sess:
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:501]: I'm uncomfortable modifying a session as a side effect when `list_checkpoints` is called. What if the session stored in `self._session` already had a stored `_http_session`?

Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:501]: With `list_checkpoints` coded in this way, is there any way to use the same HTTP connection for two successive calls?

```python
with api.login(<details>) as sess:
  persistent_det = Determined.from_session(sess)
  persistent_det.list_checkpoints()  # closes the PoolManager in persistent_det
```
